precompiles/staking/method_distribute_test.go

```diff
@@ -0,0 +1,220 @@
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(0)}...,
+		)
+
+		// Call method.
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "invalid token amount: 0")
+
+		// Unpack and check result boolean.
+		res, err := s.methodID.Outputs.Unpack(success)
+		require.NoError(t, err)
+
+		ok := res[0].(bool)
+		require.False(t, ok)
+	})
+
+	t.Run("should fail to distribute more than allowed to staking", func(t *testing.T) {
+		// Setup test.
+		s := newTestSuite(t)
+
+		// Set caller balance.
+		s.fungibleKeeper.DepositZRC20(s.ctx, s.zrc20Address, s.defaultCaller, big.NewInt(1000))
+
+		// Allow staking to spend ZRC20 tokens.
+		allowStaking(t, s, big.NewInt(999))
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(1000)}...,
+		)
+
+		// Call method.
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "invalid allowance, got 999, wanted 1000")
+
+		// Unpack and check result boolean.
+		res, err := s.methodID.Outputs.Unpack(success)
+		require.NoError(t, err)
+
+		ok := res[0].(bool)
+		require.False(t, ok)
+	})
+
+	t.Run("should fail to distribute more than user balance", func(t *testing.T) {
+		// Setup test.
+		s := newTestSuite(t)
+
+		// Set caller balance.
+		s.fungibleKeeper.DepositZRC20(s.ctx, s.zrc20Address, s.defaultCaller, big.NewInt(1000))
+
+		// Allow staking to spend ZRC20 tokens.
+		allowStaking(t, s, big.NewInt(100000))
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(1001)}...,
+		)
+
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "execution reverted")
+
+		// Unpack and check result boolean.
+		res, err := s.methodID.Outputs.Unpack(success)
+		require.NoError(t, err)
+
+		ok := res[0].(bool)
+		require.False(t, ok)
+	})
+
+	t.Run("should distribute and lock ZRC20", func(t *testing.T) {
+		// Setup test.
+		s := newTestSuite(t)
+
+		// Set caller balance.
+		s.fungibleKeeper.DepositZRC20(s.ctx, s.zrc20Address, s.defaultCaller, big.NewInt(1000))
+
+		// Allow staking to spend ZRC20 tokens.
+		allowStaking(t, s, big.NewInt(1000))
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(1000)}...,
+		)
+
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.NoError(t, err)
+
+		// Unpack and check result boolean.
+		res, err := s.methodID.Outputs.Unpack(success)
+		require.NoError(t, err)
+
+		ok := res[0].(bool)
```

Comment:
Is it possible to check the `DistributeLog` ?

Reply:
That's done during e2e in a real setup.